src/Http/WebUtilities/src/HttpResponsePipeWriter.cs

```diff
@@ -0,0 +1,275 @@
+        public override void Write(string? value)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            if (value == null)
+            {
+                return;
+            }
+
+            Write(value.AsSpan());
+        }
+
+        public override void WriteLine(ReadOnlySpan<char> value)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            Write(value);
+            Write(NewLine);
+        }
+
+        public override Task WriteAsync(char value)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            _singleCharArray[0] = value;
+
+            return WriteAsync(_singleCharArray, 0, 1);
+        }
+
+        public override Task WriteAsync(char[] values, int index, int count)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            if (values == null || count == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            var value = new Span<char>(values, index, count);
+            Write(value);
+
+            return Task.CompletedTask;
+        }
+
+        public override Task WriteAsync(string? value)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            var length = _encoder.GetByteCount(value, false);
+            var buffer = _writer.GetSpan(length);
+            _encoder.GetBytes(value, buffer, false);
+            _writer.Advance(length);
+
+            return Task.CompletedTask;
+        }
+
+        public override Task WriteAsync(ReadOnlyMemory<char> value, CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (value.IsEmpty)
+            {
+                return Task.CompletedTask;
+            }
+
+            Write(value.Span);
+
+            return Task.CompletedTask;
+        }
+
+        public override Task WriteLineAsync(ReadOnlyMemory<char> value, CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                return GetObjectDisposedTask();
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (value.IsEmpty && NewLine.Length == 0)
+            {
+                return Task.CompletedTask;
+            }
+
+            Write(value.Span);
+            Write(NewLine);
+
```

Comment:
It's a little bit of a red flag that the `WriteAsync` methods don't call `PipeWriter.FlushAsync`. This means that it's going to buffer until Flush is called by the caller even if you're using all of the async methods.

Reply:
Would it not be switching to chunking if I flush?

Reply:
Hi @rynowak , if I flush after every write it will cause sending lots of small chunks of a few bytes, as the rendering of a view involve a call to write for every element. What am I missing?

Currently it is buffered on disk so it is sent as a single chunk. I'm still not sure if that is a requirement (I believe it is given https://github.com/dotnet/aspnetcore/blob/0889a62250e56066055d45a6fce99413d3f48d56/src/Http/WebUtilities/src/HttpResponseStreamWriter.cs#L453-L454) or if we can have a limited buffer just too avoid too much chunking. Who can shed some light on this?

Thank you!